invoices/invoiceregistry_test.go

```diff
@@ -547,3 +550,84 @@ func TestUnknownInvoice(t *testing.T) {
 		t.Fatal("expected invoice not found error")
 	}
 }
+
+// Tests that invoices are canceled after expiration.
+func TestInvoiceExpiryWithRegistry(t *testing.T) {
+	t.Parallel()
+
+	cdb, cleanup, err := newTestChannelDB()
+	defer cleanup()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expiryWatcher := NewInvoiceExpiryWatcher(&testClock{})
+	registry := NewRegistry(cdb, expiryWatcher, testFinalCltvRejectDelta)
+
+	// First prefill the Channel DB with some pre-existing invoices,
+	// half of them still pending, half of them expired.
+	const numExpired = 50
+	const numPending = 50
+	existingInvoices := generateInvoiceExpiryTestData(
+		t, testNow, numExpired, numPending,
+	)
+
+	var expectedCancellations []lntypes.Hash
+
+	for paymentHash, expiredInvoice := range existingInvoices.expiredInvoices {
+		if _, err := cdb.AddInvoice(expiredInvoice, paymentHash); err != nil {
+			t.Fatalf("cannot add invoice to channel db: %v", err)
+		}
+		expectedCancellations = append(expectedCancellations, paymentHash)
+	}
+
+	for paymentHash, pendingInvoice := range existingInvoices.pendingInvoices {
+		if _, err := cdb.AddInvoice(pendingInvoice, paymentHash); err != nil {
+			t.Fatalf("cannot add invoice to channel db: %v", err)
+		}
+	}
+
+	if err = registry.Start(); err != nil {
+		t.Fatalf("cannot start registry: %v", err)
+	}
+
+	// Now generate pending and expired invoices and add them to
+	// the registry while it is up and running.
+	newInvoices := generateInvoiceExpiryTestData(
+		t, testNow, numExpired, numPending,
+	)
+
+	for paymentHash, expiredInvoice := range newInvoices.expiredInvoices {
+		_, err := registry.AddInvoice(expiredInvoice, paymentHash)
+		expectedCancellations = append(expectedCancellations, paymentHash)
+
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	for paymentHash, pendingInvoice := range newInvoices.pendingInvoices {
+		_, err := registry.AddInvoice(pendingInvoice, paymentHash)
+
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	<-time.After(testTimeout)
```

Comment:
`time.Sleep`?

Reply:
I would expect the test clock being manipulated here. That is what it is for right?

Reply:
Thanks, not sure why I use `time.After` afterall :)
In this case we must wait (time is already manipulated in the test), as `InvoiceExpiryWathcer` must go trough all invoices and cancel them through `InvoiceRegistry`. We may skip the waiting by waiting for state changes on each invoice's invoice subscriptions. I decided to instead use `LookupInvoice` (as discussed with you offline).